### PR TITLE
test(dpg): add read-only model property case

### LINKED ISSUE #2341: Generate models with read-only properties from Cadl
### Input/Output/Round-trip models with read-only properties
This issue tracks the work to implement generation of models with read-only properties in the DPG generator.

Input models are not addressed, because if a model is input-only, it should not have any read-only properties.

If we need to triage this work, generation of Round-trip models is P0, and generation of Output models is P1.

#### Overview

Read-only properties are indicated via the `@visibility("read")` decorator in Cadl, see https://github.com/microsoft/cadl/blob/main/docs/tutorial.md#visibility-decorators.

#### Input Cadl file
[Cadl testserver models/readonly-properties](https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/models/readonly-properties)

#### Expected output

OutputModel files:
- [x] [OutputModel.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/OutputModel.cs)
- [x] [OutputModel.Serialization.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/OutputModel.Serialization.cs)
- [x] Output model has no initialization constructor, unless it contains collection properties. Collection properties are initialized in an internal initialization constructor. (See discussion below in https://github.com/Azure/autorest.csharp/issues/2341#issuecomment-1220387199)
- [x] Output model has serialization constructor that takes all property values
- [x] Readonly properties are get-only
- [x] Readonly model properties are get-only
- [x] Readonly collection properties are IReadOnlyLists
- [x] DeserializeOutputModel method is defined in serialization partial class
- [x] Readonly properties are deserialized

RoundTripModel files:
- [x] [RoundTripModel.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.cs)
- [x] [RoundTripModel.Serialization.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.Serialization.cs)
- [x] Round-trip model initialization constructor does not take read-only property values
- [x] Round-trip model initialization constructor initialized read-only collection properties to ReadOnlyList
- [x] Round-trip model has serialization constructor that takes all property values
- [x] Readonly properties are get-only
- [x] Readonly model properties are get-only
- [x] Readonly collection properties are IReadOnlyLists
- [x] Write operation is defined in serialization partial class
- [x] Readonly properties are NOT serialized
- [x] DeserializeOutputModel method is defined in serialization partial class
- [x] Readonly properties are deserialized

ReadonlyModel files:
- [ ] [ReadonlyModel.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/ReadonlyModel.cs)
- [ ] [ReadonlyModel.Serialization.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/ReadonlyModel.Serialization.cs)
- [ ] ReadonlyModel is generated as an output-only model

### PR DESCRIPTION
**Pull request by archerzz**
resolve #2341

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs [hunk 1]
```diff
@@ -151,6 +163,8 @@ public partial class RoundTripModel: global::Azure.Core.IUtf8JsonSerializable
 int requiredReadonlyInt = default;
 global::Azure.Core.Optional<string> optionalReadonlyString = default;
 global::Azure.Core.Optional<int?> optionalReadonlyInt = default;
+global::Cadl.TestServer.ReadonlyProperties.Models.ReadonlyModel requiredReadonlyModel = default;
```
Comment (chunyu3) [inline test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs:166]: Are those generated codes? why do they use the type path, not use `using` to import?

File: test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs [hunk 2]
```diff
@@ -177,6 +191,18 @@ public partial class RoundTripModel: global::Azure.Core.IUtf8JsonSerializable
...
 }
+if(property.NameEquals(""requiredReadonlyModel"")){
+requiredReadonlyModel = global::Cadl.TestServer.ReadonlyProperties.Models.ReadonlyModel.DeserializeReadonlyModel(property.Value);
+continue;
+}
+if(property.NameEquals(""optionalReadonlyModel"")){
+if (property.Value.ValueKind == global::System.Text.Json.JsonValueKind.Null)
+{
+property.ThrowNonNullablePropertyIsNull();
+continue;}
```
Comment (lirenhe) [inline test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs:202]: why we need to continue after throw the exception?

File: test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs (hunk 1 above)
Comment (archerzz) [reply test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs:166]: This is not the final codes. It's the intermediate form which will be processed by Roslyn and then generate the final codes.

File: test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs (hunk 2 above)
Comment (archerzz) [reply test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/ReadonlyPropertyWritingTests.cs:202]: @lirenhe This is a debug assertion. See https://github.com/Azure/autorest.csharp/blob/795b9d22788b838f8064d40f0dd7e7d7ebe744b1/src/assets/Generator.Shared/JsonElementExtensions.cs#L94-L99

**Review by lirenhe — APPROVED** [review]

**Pull request merged**